core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1217,6 +1289,61 @@ private[spark] class TaskSetManager(
   def executorAdded(): Unit = {
     recomputeLocality()
   }
+
+  /**
+   * A class for checking inefficient tasks to be speculated, the inefficient tasks come from
+   * the tasks which may be speculated by the previous strategy.
+   */
+  private[scheduler] class InefficientTaskCalculator {
+    var taskProgressThreshold = 0.0
+    var updateSealed = false
+    private var lastComputeMs = -1L
+
+    def maybeRecompute(nowMs: Long): Unit = {
+      if (!updateSealed && (lastComputeMs <= 0 ||
+        nowMs > lastComputeMs + speculationTaskStatsCacheInterval)) {
+        var successRecords = 0L
+        var successRunTime = 0L
+        var numSuccessTasks = 0L
+        taskInfos.values.filter(_.status == "SUCCESS").foreach { taskInfo =>
+          successRecords += taskInfo.successRecords
+          successRunTime += taskInfo.successRunTime
+          numSuccessTasks += 1
+        }
```

Comment:
Note, for existing speculative execution, we simply do similar to `taskInfos.values` - that is take all successful tasks into account. That should be fine too - I am still trying to think through the implications of both.

Reply:
I wonder is it possible that we could have multiple successful attempts for a single partition? Seems like we have prevented this scenario by: 
```
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L784-L785
// Check if any other attempt succeeded before this and this attempt has not been handled
if (successful(index) && killedByOtherAttempt.contains(tid)) {
```

And even if it's possible, I think it's still fine since we only calculate the average number here.